crates/video/src/main.rs

```diff
@@ -0,0 +1,213 @@
+                    "vhost-user connection closed with partial message.
+                    If the VM is shutting down, this is expected behavior;
+                    otherwise, it might be a bug."
+                );
+            }
+            Err(e) => {
+                warn!("Error running daemon: {:?} -> {}", e, e.to_string());
+            }
+        }
+
+        vu_video_backend
+            .read()
+            .unwrap()
+            .exit_event
+            .write(1)
+            .expect("Shutting down worker thread");
+    }
+}
+
+fn main() -> Result<()> {
+    env_logger::init();
+
+    start_backend(VuVideoConfig::try_from(VideoArgs::parse()).unwrap())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use assert_matches::assert_matches;
+    use rstest::*;
+    use tempfile::tempdir;
+
+    #[rstest]
+    // No device specified defaults to /dev/video0
+    #[case::no_device(vec!["", "-s  video.sock ", "-b", "null"],
+    VideoArgs {
+        socket_path: "video.sock".to_string(),
+        v4l2_device: "/dev/video0".to_string(),
+        backend: BackendType::Null,
+    })]
+    // Specifying device overwrite the default value
+    #[case::set_device(vec!["", "-s  video.sock ", "-d /dev/video1 ", "-b", "null"],
+    VideoArgs {
+        socket_path: "video.sock".to_string(),
+        v4l2_device: "/dev/video1".to_string(),
+        backend: BackendType::Null,
+    })]
+    // Selecting different decoder
+    #[case::set_v4l2_decoder(vec![" ", "--socket-path", "long-video.sock ", "-b", "v4l2-decoder"],
+    VideoArgs {
+        socket_path: "long-video.sock".to_string(),
+        v4l2_device: "/dev/video0".to_string(),
+        backend: BackendType::V4L2Decoder,
+    })]
+    fn test_command_line_arguments(#[case] args: Vec<&str>, #[case] command_line: VideoArgs) {
+        let args: VideoArgs = Parser::parse_from(args.as_slice());
+
+        assert_eq!(
+            VuVideoConfig::try_from(command_line).unwrap(),
+            VuVideoConfig::try_from(args).unwrap()
+        );
+    }
+
+    #[test]
+    fn test_fail_create_backend() {
+        use vhu_video::VuVideoError;
+        let config = VideoArgs {
+            socket_path: "video.sock".to_string(),
+            v4l2_device: "/path/invalid/video.dev".to_string(),
+            backend: BackendType::V4L2Decoder,
+        };
+        assert_matches!(
+            start_backend(VuVideoConfig::try_from(config.clone()).unwrap()).unwrap_err(),
+            Error::CouldNotCreateBackend(VuVideoError::AccessVideoDeviceFile)
+        );
+    }
+
+    #[test]
+    fn test_fail_listener() {
+        use std::fs::File;
+        let test_dir = tempdir().expect("Could not create a temp test directory.");
+        let v4l2_device = test_dir.path().join("video.dev").display().to_string();
+        File::create(v4l2_device.clone()).expect("Could not create a test device file.");
+        let config = VideoArgs {
+            socket_path: "~/path/invalid/video.sock".to_string(),
```

Comment:
```suggestion
            socket_path: "~/path/invalid/video.sock".to_owned(),
```

Reply:
This should not matter anymore in modern Rust. Personally, I would just use `.into()`. They are all equal performance wise.

Reply:
It is not about performance here but semantics. `.into()` is the worse from readability point of view.

Reply:
Hm. But wouldn't the initial `.to_string()` be the most readable then? :innocent::thinking:
But I got no strong feelings here.

Reply:
ToString is meant for converting a type that is not a string to it string representation. Where here you go from an unowned string (str) to an owned one (String)

Reply:
Anyway, let's just turn it into a Path/PathBuf here :P